src/liballoc/collections/vec_deque.rs

```diff
@@ -202,6 +202,30 @@ impl<T> VecDeque<T> {
                                  len);
     }
 
+    /// Copies all values from `src` to the back of `self`, wrapping around if needed.
+    ///
+    /// # Safety
+    ///
+    /// The capacity must be sufficient to hold self.len() + src.len() elements.
+    /// If so, this function never panics.
+    #[inline]
+    unsafe fn copy_slice(&mut self, src: &[T]) {
+        let expected_new_len = self.len() + src.len();
+        debug_assert!(self.capacity() >= expected_new_len);
+
+        let dst_high_ptr = self.ptr().add(self.head);
```

Comment:
For this to not invoke UB:

* `self.head` has to be smaller than `isize` here, 
*  `ptr` + `head` cannot overflow an `usize`

I think we need either a comment stating why this two conditions always hold (and maybe a debug assert for them).

Reply:
_I think_ `RawVec` only allows allocations up to `isize::max_value()`, which means that `head` and `tail` will always be smaller than `isize`. Also, if the allocation succeeded, then `ptr+head` cannot overflow `usize`, so this operation is always safe.

Basically, one could `debug_assert!(self.capacity() <= isize::max_value() as usize)` here, but I don't know if this is worth it. cc @alexcrichton ?

Reply:
@MaloJaffre I think I'll prefer to document this with a `debug_assert!` because we might change `RawVec` in the future (its unstable) and that could end up silently breaking `VecDeque`